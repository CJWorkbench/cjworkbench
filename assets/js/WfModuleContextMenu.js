// Drop-down menu on Workflows List page, for each listed WF
// triggered by click on three-dot icon next to listed workflow module

import React from 'react'
import {
    UncontrolledDropdown,
    DropdownToggle,
    DropdownMenu,
    DropdownItem,
    Modal,
    ModalHeader,
    ModalBody,
    ModalFooter,
    Form,
    FormGroup,
    Label,
    Input
  } from 'reactstrap'
import PropTypes from 'prop-types'
import CopyToClipboard from 'react-copy-to-clipboard';


export default class WfModuleContextMenu extends React.Component {
  constructor(props) {
    super(props);
    this.deleteOption = this.deleteOption.bind(this);
    this.toggleExportModal = this.toggleExportModal.bind(this);
    this.renderExportModal = this.renderExportModal.bind(this);
    this.onCsvCopy = this.onCsvCopy.bind(this);
    this.onCsvLeave = this.onCsvLeave.bind(this);
    this.onJsonCopy = this.onJsonCopy.bind(this);
    this.onJsonLeave = this.onJsonLeave.bind(this);
    this.state = {
      exportModalOpen: false,
      csvCopied: false,
      jsonCopied: false
    };
  }

  deleteOption() {
    this.props.removeModule();
  }

  toggleExportModal() {
    this.setState({ exportModalOpen: !this.state.exportModalOpen });
  }

  csvUrlString(id) {
    var path = "/public/moduledata/live/" + id + ".csv";
    // allowing an out for testing (there is no window.location.href during test)
    if (window.location.href == 'about:blank') {
      return path;
    } else {
      var url = new URL(path, window.location.href).href;
      return url;
    }
  }

  jsonUrlString(id) {
    var path = "/public/moduledata/live/" + id + ".json";
    // allowing an out for testing (there is no window.location.href during test)
    if (window.location.href == 'about:blank') {
      return path;
    } else {
      var url = new URL(path, window.location.href).href;
      return url;
    }
  }

  onCsvCopy() {
    this.setState({csvCopied: true});
  }

  onCsvLeave() {
    this.setState({csvCopied: false});
  }

  onJsonCopy() {
    this.setState({jsonCopied: true});
  }

  onJsonLeave() {
    this.setState({jsonCopied: false});
  }

  renderCsvCopyLink() {
    var csvString = this.csvUrlString(this.props.id);

    if (this.state.csvCopied) {
      return (
        <div className='info-2 t-orange mt-3' onMouseLeave={this.onCsvLeave}>CSV link copied to clipboard</div>
      );
    } else {
      return (
        <CopyToClipboard text={csvString} onCopy={this.onCsvCopy} className='info-2 t-f-blue mt-3 test-csv-copy'>
          <div>Copy live link</div>
        </CopyToClipboard>
      );
    }
  }

  renderJsonCopyLink() {
    var jsonString = this.jsonUrlString(this.props.id);

    if (this.state.jsonCopied) {
      return (
        <div className='info-2 t-orange mt-3' onMouseLeave={this.onJsonLeave}>JSON link copied to clipboard</div>
      );
    } else {
      return (
        <CopyToClipboard text={jsonString} onCopy={this.onJsonCopy} className='info-2 t-f-blue mt-3 test-json-copy'>
          <div>Copy live link</div>
        </CopyToClipboard>
      );
    }
  }

  renderExportModal() {
    if (!this.state.exportModalOpen) {
      return null;
    }

    var csvString = this.csvUrlString(this.props.id);
    var jsonString = this.jsonUrlString(this.props.id);
    var csvCopyLink = this.renderCsvCopyLink();
    var jsonCopyLink = this.renderJsonCopyLink();

    return (
      <Modal isOpen={this.state.exportModalOpen} toggle={this.toggleExportModal} className={this.props.className}>
        <ModalHeader toggle={this.toggleModal} className='dialog-header modal-header' >
          <div className='t-d-gray title-4'>Export Data</div>
          <div className='icon-close' onClick={this.toggleExportModal}></div>
        </ModalHeader>
        <ModalBody className='dialog-body'>
          <FormGroup>
            <div className='d-flex justify-content-between flex-row'>
              <Label className='t-d-gray info-1'>CSV</Label>
              {csvCopyLink}
            </div>
            <div className='d-flex justify-content-between flex-row mb-3'>
              <Input type='url' className='text-field t-d-gray content-2 test-csv-field' placeholder={csvString} readOnly/>
              <div className='download-icon-box'>
                <a href={csvString} className='icon-download button-icon test-csv-download' download></a>
              </div>
            </div>
            <div className='d-flex justify-content-between flex-row'>
              <Label className='t-d-gray info-1'>JSON</Label>
              {jsonCopyLink}
            </div>
            <div className='d-flex justify-content-between flex-row'>
              <Input type='url' className='text-field t-d-gray content-2 test-json-field' placeholder={jsonString} readOnly/>
              <div className='download-icon-box'>
                <a href={jsonString} className='icon-download button-icon test-json-download' download></a>
              </div>
            </div>
          </FormGroup>
        </ModalBody>
        <ModalFooter className='dialog-footer'>
          <div onClick={this.toggleExportModal} className='button-blue action-button test-done-button'>Done</div>{' '}
        </ModalFooter>
      </Modal>
    );
  }

  render() {
    var exportModal = this.renderExportModal();

    return (
       <UncontrolledDropdown onClick={this.props.stopProp}>
        <DropdownToggle className='context-button'>
          <div className='module-menu-icon icon-more'></div>
        </DropdownToggle>
        <DropdownMenu right className='dropdown-menu'>
          {/* Opens Modal window for downloading files */}
          <DropdownItem key={1} onClick={this.toggleExportModal} className='dropdown-menu-item mb-1 test-export-button'>
            <span className='icon-download button-icon'></span>
            <span className='t-d-gray content-3 ml-3'>Export</span>
            {exportModal}
          </DropdownItem>
          {/* Currently does nothing */}
          <DropdownItem key={2} className='dropdown-menu-item test-update-button'>
            <span className='icon-info button-icon'></span>
            <span className='t-d-gray content-3 ml-3'>Update</span>
          </DropdownItem>
          {/* Will delete the parent WF Module from the list */}
          <DropdownItem key={3} onClick={this.deleteOption} className='dropdown-menu-item test-delete-button'>
            <span className='icon-bin button-icon'></span>
            <span className='t-d-gray content-3 ml-3'>Delete</span>
          </DropdownItem>
        </DropdownMenu>
       </UncontrolledDropdown>
    );
  }
}

WfModuleContextMenu.propTypes = {
  removeModule: PropTypes.func,
  id:           PropTypes.number,
  stopProp:     PropTypes.func
};
