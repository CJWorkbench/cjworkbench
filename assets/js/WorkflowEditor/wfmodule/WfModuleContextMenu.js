// Drop-down menu at upper right of each module in a workflow
import React from 'react'
import PropTypes from 'prop-types'
import { UncontrolledDropdown, DropdownToggle, DropdownMenu, DropdownItem } from '../../components/Dropdown'
import ExportModal from '../../ExportModal'
import { Trans, t } from '@lingui/macro'
import { withI18n } from '@lingui/react'

const WfModuleContextMenu = React.memo(function WfModuleContextMenu ({ i18n, removeModule, id }) {
  const [isExportModalOpen, setExportModalOpen] = React.useState(false)
  const handleClickOpenExportModal = React.useCallback(() => setExportModalOpen(true))
  const handleCloseExportModal = React.useCallback(() => setExportModalOpen(false))
  const handleClickDelete = React.useCallback(() => removeModule(id))

  return (
    <UncontrolledDropdown>
      <DropdownToggle title={i18n._(t('js.WorkflowEditor.wfmodule.WfModuleContextMenu.more')`more`)} className='context-button'>
        <i className='icon-more' />
      </DropdownToggle>
      <DropdownMenu>
        <DropdownItem onClick={handleClickOpenExportModal} className='test-export-button' icon='icon-download'><Trans id='js.WorkflowEditor.wfmodule.wfModuleContextMenu.exportData'>Export data</Trans></DropdownItem>
        <DropdownItem onClick={handleClickDelete} className='test-delete-button' icon='icon-bin'><Trans id='js.WorkflowEditor.wfmodule.wfModuleContextMenu.delete'>Delete</Trans></DropdownItem>
      </DropdownMenu>
      <ExportModal open={isExportModalOpen} wfModuleId={id} toggle={handleCloseExportModal} />
    </UncontrolledDropdown>
  )
})
WfModuleContextMenu.propTypes = {
  removeModule: PropTypes.func,
  id: PropTypes.number
}
export default withI18n()(WfModuleContextMenu)
