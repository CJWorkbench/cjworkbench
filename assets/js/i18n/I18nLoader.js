import React from 'react'
import { I18nProvider } from '@lingui/react'
import fetchCatalog from './catalogs'

export class I18nLoader extends React.Component {
  state = {
    catalogs: {},
  }
  
  loadCatalog = (language) => {
      this.setState(state => ({
          catalogs: {
            ...state.catalogs,
            [language]: fetchCatalog(language)
          }
      }))
  }
  
  componentDidMount() {
    this.loadCatalog(this.props.language)
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { language } = nextProps
    const { catalogs } = nextState

    if (language !== this.props.language && !catalogs[language]) {
      this.loadCatalog(language)
      return false
    }

    return true
  }

  render () {
    const { children, language } = this.props
    const { catalogs } = this.state

    // Skip rendering when catalog isn't loaded.
    if (!catalogs[language]) {
        return null
    }

    return (
      <I18nProvider language={language} catalogs={catalogs}>
        {children}
      </I18nProvider>
    )
  }
}
