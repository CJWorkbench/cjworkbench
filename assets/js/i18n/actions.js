const SET_LOCALE = 'SET_LOCALE';

function reduceSetLocale(state, action){
  return {
    ...state,
    locale: action.locale
  };
}

export function setLocaleAction(locale){
  return {
    type: SET_LOCALE,
    locale: locale
  }
}

export const reducerFunctions = {
  [SET_LOCALE]: reduceSetLocale
}

export function localeReducer(state, action){
  if (action.type in reducerFunctions) {
    // Run a registered reducer
    return reducerFunctions[action.type](state, action)
  }
  
  return state;
}
