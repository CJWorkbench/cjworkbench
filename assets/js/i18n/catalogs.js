import enCatalog from '../../locale/en/messages.po'
import elCatalog from '../../locale/el/messages.po'

const catalogs = {
  en: enCatalog,
  el: elCatalog
}

export default function fetchCatalog (language) {
  return catalogs[language]
}
