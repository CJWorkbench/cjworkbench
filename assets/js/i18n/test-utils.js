import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
