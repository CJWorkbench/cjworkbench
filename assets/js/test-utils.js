/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
