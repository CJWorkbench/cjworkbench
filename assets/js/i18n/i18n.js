import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
