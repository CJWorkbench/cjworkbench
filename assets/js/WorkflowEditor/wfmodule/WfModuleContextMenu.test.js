/* globals afterEach, beforeEach, describe, expect, it, jest */
import React from 'react'
import WfModuleContextMenu from './WfModuleContextMenu'
import { mountWithI18n } from '../../i18n/test-utils'

describe('WfModuleContextMenu', () => {
  let wrapper
  let removeModule

  beforeEach(() => {
    removeModule = jest.fn()

    wrapper = mountWithI18n(
      <WfModuleContextMenu
        removeModule={removeModule}
        id={415}
        className='menu-test-class'
      />
    )
  })

  afterEach(() => wrapper.unmount())

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  // only checking the call to removeModule(), not the removal
  it('Renders menu option to Delete with onClick method', () => {
    // open the context menu
    wrapper.find('button.context-button').simulate('click')

    const deleteButton = wrapper.find('button.test-delete-button')
    expect(deleteButton).toHaveLength(1)
    deleteButton.simulate('click')
    expect(removeModule).toHaveBeenCalled()
  })

  it('should open and close the export modal', () => {
    // open the context menu
    wrapper.find('button.context-button').simulate('click')

    // open the modal window
    const exportButton = wrapper.find('button.test-export-button')
    exportButton.simulate('click')
    expect(wrapper.find('div.modal-dialog')).toHaveLength(1)

    // close it
    wrapper.find('div.modal-dialog button.close').simulate('click')
    wrapper.update()
    expect(wrapper.find('div.modal-dialog')).toHaveLength(0)
  })
})
