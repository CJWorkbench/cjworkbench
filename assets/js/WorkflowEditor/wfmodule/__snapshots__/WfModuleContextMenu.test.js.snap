// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WfModuleContextMenu should match snapshot 1`] = `
<WithI18n
  className="menu-test-class"
  i18n={
    Object {
      "_": [Function],
      "_activeLanguageData": Object {
        "plurals": [Function],
      },
      "_activeMessages": Object {},
      "_catalogs": Object {
        "en": Object {
          "languageData": Object {},
          "messages": Object {},
        },
      },
      "_dev": Object {
        "compile": [Function],
        "loadLanguageData": [Function],
      },
      "_language": "en",
      "_locales": undefined,
      "_missing": undefined,
      "plural": [Function],
      "select": [Function],
      "selectOrdinal": [Function],
      "t": [Function],
    }
  }
  id={415}
  removeModule={[MockFunction]}
>
  <I18n
    update={true}
    withHash={true}
  >
    <Memo(WfModuleContextMenu)
      className="menu-test-class"
      i18n={
        Object {
          "_": [Function],
          "_activeLanguageData": Object {
            "plurals": [Function],
          },
          "_activeMessages": Object {},
          "_catalogs": Object {
            "en": Object {
              "languageData": Object {},
              "messages": Object {},
            },
          },
          "_dev": Object {
            "compile": [Function],
            "loadLanguageData": [Function],
          },
          "_language": "en",
          "_locales": undefined,
          "_missing": undefined,
          "plural": [Function],
          "select": [Function],
          "selectOrdinal": [Function],
          "t": [Function],
        }
      }
      id={415}
      removeModule={[MockFunction]}
    >
      <UncontrolledDropdown>
        <Dropdown
          isOpen={false}
          toggle={[Function]}
        >
          <Manager
            tag={false}
          >
            <div
              className="dropdown"
            >
              <DropdownToggle
                className="context-button"
                title="more"
              >
                <Reference>
                  <InnerReference
                    setReferenceNode={[Function]}
                  >
                    <button
                      aria-expanded={false}
                      className="btn btn-secondary context-button"
                      disabled={false}
                      onClick={[Function]}
                      title="more"
                    >
                      <i
                        className="icon-more"
                      />
                    </button>
                  </InnerReference>
                </Reference>
              </DropdownToggle>
              <DropdownMenu />
              <ExportModal
                open={false}
                toggle={[Function]}
                wfModuleId={415}
              >
                <Modal
                  isOpen={false}
                  toggle={[Function]}
                />
              </ExportModal>
            </div>
          </Manager>
        </Dropdown>
      </UncontrolledDropdown>
    </Memo(WfModuleContextMenu)>
  </I18n>
</WithI18n>
`;
